light-client/src/fork_detector.rs

```diff
@@ -0,0 +1,88 @@
+use serde::{Deserialize, Serialize};
+
+use crate::prelude::*;
+use crate::supervisor::Instance;
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum ForkDetection {
+    Detected(Vec<Fork>),
+    NotDetected,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum Fork {
+    Forked(LightBlock),
+    Faulty(LightBlock),
+}
+
+pub trait ForkDetector {
+    fn detect_forks(
+        &self,
+        light_block: &LightBlock,
+        trusted_state: &LightBlock,
+        secondaries: Vec<&Instance>,
+    ) -> ForkDetection;
+}
+
+pub struct ProdForkDetector {}
+
+impl ProdForkDetector {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for ProdForkDetector {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ForkDetector for ProdForkDetector {
+    fn detect_forks(
+        &self,
+        light_block: &LightBlock,
+        trusted_state: &LightBlock,
+        secondaries: Vec<&Instance>,
+    ) -> ForkDetection {
+        let mut forks = Vec::with_capacity(secondaries.len());
+
+        for secondary in secondaries {
+            let mut state = State::new(MemoryStore::new());
+
+            let secondary_block = secondary
+                .light_client
+                .get_or_fetch_block(light_block.height(), &mut state)
+                .unwrap(); // FIXME: unwrap
+
+            // TODO: Should hash the headers here instead of comparing them for equality?
```

Comment:
I think that would be the right thing to do, yes! (theoretically comparing the headers vs comparing their hashes should be equivalent, practically that might differ although I don't think that can be the case here - still using the hashes would be consistent imo).